torch/nn/modules/linear.py

```diff
@@ -178,4 +179,78 @@ def extra_repr(self) -> str:
         )
 
+
+class LazyLinear(LazyModuleMixin, Linear):
+    r"""A :class:`torch.nn.Linear` module with lazy initialization.
+
+    This module will initialize the `weight` and `bias` parameters after
+    the first call to ``forward`` is done. After that the module will
+    become a :class:`torch.nn.Linear` dropping all the lazy initialization
+    logic.
+
+    Be sure to run "dummy" ``forward`` once to initialize all parameters in case
+    that you need to access the weights before the actual "real" ``forward`` pass.
+
+    Note that lazy modules cannot validate if the shape is correct during
+    deserialization.  Also note that the initial weights values may become different
+    from the original (non-lazy) module even if the random seed is manually
+    configured, as the order of initialization is different from the original
+    one; especially, ``module.cuda()`` may cause the initialization to run on
+    a GPU.
+    The default value of lazy parameters is ``UninitializedParameter()``.
+
+    Args:
+        out_features: size of each output sample
+        bias: If set to ``False``, the layer will not learn an additive bias.
+            Default: ``True``
+
+    Attributes:
+        weight: the learnable weights of the module of shape
+            :math:`(\text{out\_features}, \text{in\_features})`. The values are
+            initialized from :math:`\mathcal{U}(-\sqrt{k}, \sqrt{k})`, where
+            :math:`k = \frac{1}{\text{in\_features}}`
+        bias:   the learnable bias of the module of shape :math:`(\text{out\_features})`.
+                If :attr:`bias` is ``True``, the values are initialized from
+                :math:`\mathcal{U}(-\sqrt{k}, \sqrt{k})` where
+                :math:`k = \frac{1}{\text{in\_features}}`
+
+    .. note:: A `LazyModule`, or a module containing a `LazyModule` requires an explicit
+    dummy forward call in the following scenarios.
+
+    * Access an :class:`UnitializedParameter` data or properties
+    >>> lazy_module.weight.shape
+
+    * Use an Optimizer that accesses Parameters data in their constructor 
+      (Currently only :class:`torch.optim.AdaGrad`) other optimizers work without
+      explicitly initializing the module
+
+    >>> optim = torch.optim.AdaGrad(lazy_module.parameters())
+
+    * Use :class:`torch.nn.DataParallel` or :class:`torch.nn.DistributedDataParallel`
+    >>> module_dp = torch.nn.DataParallel(lazy_module)
+
+    * Module reparametrization using :func:`torch.nn.utils.weight_norm` or :func:`torch.nn.utils.spectral_norm`
+    >>> torch.nn.utils.weight_norm(lazy_module)
+
+    * When dealing with non-default module initialization
+    >>> torch.nn.init.uniform_(lazy_module.weight)
+
+    """
+
+    cls_to_become = Linear
+
+    def __init__(self, out_features: int, bias: bool = True) -> None:
+        super().__init__(0, out_features, bias)
+        self.weight = UninitializedParameter()    
+
+    def reset_parameters(self) -> None:
```

Comment:
I'm curious if we can make it so that:

- attempts to access the parameters of a lazy module throws an error
- the forward pass has the logic to "materialize" AND register the params AND convert the module to not be lazy

Reply:
> attempts to access the parameters of a lazy module throws an error
When initializing the parameters we need to explicitly access them, I think it's not easy at all to discriminate whether do we have or not legit accesses to a parameter.

> the forward pass has the logic to "materialize" AND register the params AND convert the module to not be lazy

Sorry, can you provide some kind of example? I don't currently see how this could be done.

Reply:
I understand. I'm asking what if we broke the property that references to an UninitializedParameter are supported.

For example, in linear we have weight and bias. Could we do something like the following:

- Delete the weight and bias parameters and make attempts to access them throw an error
- Create dummy _weight and _bias UninitializedParameters that are just empty tensors that can have their dtype and device change, but trying to edit them otherwise is an error
- Have the LazyModule mixin cause attempts to access the parameters to throw an error
- Then, on forward (or pre forward), the weight and bias parameters are created as usual with shapes derived from the input data and dtype and device derived from _weight and _bias
- The _weight and _bias UninitializedParameters are removed
- The LazyModule class is removed

I think that would give us a regular linear layer? Put another way, I'm wonder if we can create a module that will let the network it's in work with .to(), but it requires a forward pass before anything else can be done to it (like registering an optimizer).

Reply:
This is quite a design change with several implications,
Right now an `UninitializedParameter` changes it's data and `__class__` when it is materialized.
https://github.com/pytorch/pytorch/blob/6de9ebd676bed3ff3ed269767a6a8a2ab2a503a2/torch/nn/parameter.py#L96-L97

But for the materialization to happen, we need to access the parameter itself, so we can't forbid this access in the current pattern.
https://github.com/pytorch/pytorch/blob/6de9ebd676bed3ff3ed269767a6a8a2ab2a503a2/torch/nn/modules/linear.py#L254

With this trick, if some code such as optimizers were holding a reference to an `UninitializedParameter`, that reference will still be valid after the materialization. However, this was a trick to avoid the explicit dry run in simple cases, so your proposal actually makes the design more robust if we forbid accessing a state dict and make the dry run mandatory.
I would like to hear @albanD and @yf225 opinions on this. I am fine with either of the approaches as they have both their cons/pros.

Reply:
Great points. See my comment below, too. Maybe we wait on this decision. But let me check with the team.

Reply:
Per discussion elsewhere, let's not make this change now. We'll revisit it later.